src/dbginfo is hopelessly broken

src/dbginfo doesn't compile "as is", with or without -I../common

bare minimum, this needs a working Makefile
it should also be included in the build.

if it is intended to work some other way than what i have here,
then that also needs to be documented,
and additional code / Makefiles should be included to demonstrate.

if we're NOT going to give that minimum level of support,
it should simply be removed altogether.

<pre>
gorilla@guest:~/cc65/src/dbginfo (master)$ gcc -g  *.c -o dbgsh
dbgsh.c:44:10: fatal error: attrib.h: No such file or directory
   44 | #include "attrib.h"
      |          ^~~~~~~~~~
compilation terminated.
</pre>

<pre>
gorilla@guest:~/cc65/src/dbginfo (master)$ gcc -g -I../common *.c -o dbgsh
dbgsh.c:60:22: error: unknown type name ‘Collection’
   60 | static void CmdHelp (Collection* Args attribute ((unused)));
      |                      ^~~~~~~~~~
dbgsh.c:63:22: error: unknown type name ‘Collection’
   63 | static void CmdLoad (Collection* Args);
      |                      ^~~~~~~~~~
dbgsh.c:66:22: error: unknown type name ‘Collection’
   66 | static void CmdQuit (Collection* Args attribute ((unused)));
      |                      ^~~~~~~~~~
dbgsh.c:69:22: error: unknown type name ‘Collection’
   69 | static void CmdShow (Collection* Args);
      |                      ^~~~~~~~~~
dbgsh.c:72:26: error: unknown type name ‘Collection’
   72 | static void CmdShowHelp (Collection* Args);
      |                          ^~~~~~~~~~
dbgsh.c:75:33: error: unknown type name ‘Collection’
   75 | static void CmdShowChildScopes (Collection* Args);
      |                                 ^~~~~~~~~~
dbgsh.c:78:29: error: unknown type name ‘Collection’
   78 | static void CmdShowCSymbol (Collection* Args);
      |                             ^~~~~~~~~~
dbgsh.c:81:30: error: unknown type name ‘Collection’
   81 | static void CmdShowFunction (Collection* Args);
      |                              ^~~~~~~~~~
dbgsh.c:84:29: error: unknown type name ‘Collection’
   84 | static void CmdShowLibrary (Collection* Args);
      |                             ^~~~~~~~~~
dbgsh.c:87:26: error: unknown type name ‘Collection’
   87 | static void CmdShowLine (Collection* Args);
      |                          ^~~~~~~~~~
dbgsh.c:90:28: error: unknown type name ‘Collection’
   90 | static void CmdShowModule (Collection* Args);
      |                            ^~~~~~~~~~
dbgsh.c:93:27: error: unknown type name ‘Collection’
   93 | static void CmdShowScope (Collection* Args);
      |                           ^~~~~~~~~~
dbgsh.c:96:29: error: unknown type name ‘Collection’
   96 | static void CmdShowSegment (Collection* Args);
      |                             ^~~~~~~~~~
dbgsh.c:99:28: error: unknown type name ‘Collection’
   99 | static void CmdShowSource (Collection* Args);
      |                            ^~~~~~~~~~
dbgsh.c:102:26: error: unknown type name ‘Collection’
  102 | static void CmdShowSpan (Collection* Args);
      |                          ^~~~~~~~~~
dbgsh.c:105:28: error: unknown type name ‘Collection’
  105 | static void CmdShowSymbol (Collection* Args);
      |                            ^~~~~~~~~~
dbgsh.c:108:28: error: unknown type name ‘Collection’
  108 | static void CmdShowSymDef (Collection* Args);
      |                            ^~~~~~~~~~
dbgsh.c:111:28: error: unknown type name ‘Collection’
  111 | static void CmdShowSymRef (Collection* Args);
      |                            ^~~~~~~~~~
dbgsh.c:114:26: error: unknown type name ‘Collection’
  114 | static void CmdShowType (Collection* Args);
      |                          ^~~~~~~~~~
dbgsh.c:117:24: error: unknown type name ‘Collection’
  117 | static void CmdUnload (Collection* Args attribute ((unused)));
      |                        ^~~~~~~~~~
dbgsh.c:159:26: error: unknown type name ‘Collection’
  159 |     void        (*Func) (Collection*);
      |                          ^~~~~~~~~~
dbgsh.c:160:1: warning: no semicolon at end of struct or union
  160 | };
      | ^
dbgsh.c:168:9: error: ‘CmdQuit’ undeclared here (not in a function)
  168 |         CmdQuit
      |         ^~~~~~~
dbgsh.c:168:9: warning: excess elements in struct initializer
dbgsh.c:168:9: note: (near initialization for ‘MainCmds[0]’)
dbgsh.c:173:9: error: ‘CmdHelp’ undeclared here (not in a function)
  173 |         CmdHelp
      |         ^~~~~~~
dbgsh.c:173:9: warning: excess elements in struct initializer
dbgsh.c:173:9: note: (near initialization for ‘MainCmds[1]’)
dbgsh.c:178:9: error: ‘CmdLoad’ undeclared here (not in a function)
  178 |         CmdLoad
      |         ^~~~~~~
dbgsh.c:178:9: warning: excess elements in struct initializer
dbgsh.c:178:9: note: (near initialization for ‘MainCmds[2]’)
dbgsh.c:183:9: warning: excess elements in struct initializer
  183 |         CmdQuit
      |         ^~~~~~~
dbgsh.c:183:9: note: (near initialization for ‘MainCmds[3]’)
dbgsh.c:188:9: error: ‘CmdShow’ undeclared here (not in a function)
  188 |         CmdShow
      |         ^~~~~~~
dbgsh.c:188:9: warning: excess elements in struct initializer
dbgsh.c:188:9: note: (near initialization for ‘MainCmds[4]’)
dbgsh.c:193:9: error: ‘CmdUnload’ undeclared here (not in a function)
  193 |         CmdUnload
      |         ^~~~~~~~~
dbgsh.c:193:9: warning: excess elements in struct initializer
dbgsh.c:193:9: note: (near initialization for ‘MainCmds[5]’)
dbgsh.c:203:9: error: ‘CmdShowChildScopes’ undeclared here (not in a function)
  203 |         CmdShowChildScopes
      |         ^~~~~~~~~~~~~~~~~~
dbgsh.c:203:9: warning: excess elements in struct initializer
dbgsh.c:203:9: note: (near initialization for ‘ShowCmds[0]’)
dbgsh.c:208:9: error: ‘CmdShowCSymbol’ undeclared here (not in a function)
  208 |         CmdShowCSymbol
      |         ^~~~~~~~~~~~~~
dbgsh.c:208:9: warning: excess elements in struct initializer
dbgsh.c:208:9: note: (near initialization for ‘ShowCmds[1]’)
dbgsh.c:213:9: warning: excess elements in struct initializer
  213 |         CmdShowCSymbol
      |         ^~~~~~~~~~~~~~
dbgsh.c:213:9: note: (near initialization for ‘ShowCmds[2]’)
dbgsh.c:218:9: error: ‘CmdShowFunction’ undeclared here (not in a function)
  218 |         CmdShowFunction
      |         ^~~~~~~~~~~~~~~
dbgsh.c:218:9: warning: excess elements in struct initializer
dbgsh.c:218:9: note: (near initialization for ‘ShowCmds[3]’)
dbgsh.c:223:9: warning: excess elements in struct initializer
  223 |         CmdShowFunction
      |         ^~~~~~~~~~~~~~~
dbgsh.c:223:9: note: (near initialization for ‘ShowCmds[4]’)
dbgsh.c:228:9: error: ‘CmdShowHelp’ undeclared here (not in a function)
  228 |         CmdShowHelp
      |         ^~~~~~~~~~~
dbgsh.c:228:9: warning: excess elements in struct initializer
dbgsh.c:228:9: note: (near initialization for ‘ShowCmds[5]’)
dbgsh.c:233:9: error: ‘CmdShowLine’ undeclared here (not in a function)
  233 |         CmdShowLine
      |         ^~~~~~~~~~~
dbgsh.c:233:9: warning: excess elements in struct initializer
dbgsh.c:233:9: note: (near initialization for ‘ShowCmds[6]’)
dbgsh.c:238:9: error: ‘CmdShowLibrary’ undeclared here (not in a function)
  238 |         CmdShowLibrary
      |         ^~~~~~~~~~~~~~
dbgsh.c:238:9: warning: excess elements in struct initializer
dbgsh.c:238:9: note: (near initialization for ‘ShowCmds[7]’)
dbgsh.c:243:9: error: ‘CmdShowModule’ undeclared here (not in a function)
  243 |         CmdShowModule
      |         ^~~~~~~~~~~~~
dbgsh.c:243:9: warning: excess elements in struct initializer
dbgsh.c:243:9: note: (near initialization for ‘ShowCmds[8]’)
dbgsh.c:248:9: error: ‘CmdShowScope’ undeclared here (not in a function)
  248 |         CmdShowScope
      |         ^~~~~~~~~~~~
dbgsh.c:248:9: warning: excess elements in struct initializer
dbgsh.c:248:9: note: (near initialization for ‘ShowCmds[9]’)
dbgsh.c:253:9: error: ‘CmdShowSegment’ undeclared here (not in a function)
  253 |         CmdShowSegment
      |         ^~~~~~~~~~~~~~
dbgsh.c:253:9: warning: excess elements in struct initializer
dbgsh.c:253:9: note: (near initialization for ‘ShowCmds[10]’)
dbgsh.c:258:9: error: ‘CmdShowSource’ undeclared here (not in a function)
  258 |         CmdShowSource
      |         ^~~~~~~~~~~~~
dbgsh.c:258:9: warning: excess elements in struct initializer
dbgsh.c:258:9: note: (near initialization for ‘ShowCmds[11]’)
dbgsh.c:263:9: error: ‘CmdShowSpan’ undeclared here (not in a function)
  263 |         CmdShowSpan
      |         ^~~~~~~~~~~
dbgsh.c:263:9: warning: excess elements in struct initializer
dbgsh.c:263:9: note: (near initialization for ‘ShowCmds[12]’)
dbgsh.c:268:9: error: ‘CmdShowSymbol’ undeclared here (not in a function)
  268 |         CmdShowSymbol
      |         ^~~~~~~~~~~~~
dbgsh.c:268:9: warning: excess elements in struct initializer
dbgsh.c:268:9: note: (near initialization for ‘ShowCmds[13]’)
dbgsh.c:273:9: error: ‘CmdShowSymDef’ undeclared here (not in a function)
  273 |         CmdShowSymDef
      |         ^~~~~~~~~~~~~
dbgsh.c:273:9: warning: excess elements in struct initializer
dbgsh.c:273:9: note: (near initialization for ‘ShowCmds[14]’)
dbgsh.c:278:9: error: ‘CmdShowSymRef’ undeclared here (not in a function)
  278 |         CmdShowSymRef
      |         ^~~~~~~~~~~~~
dbgsh.c:278:9: warning: excess elements in struct initializer
dbgsh.c:278:9: note: (near initialization for ‘ShowCmds[15]’)
dbgsh.c:283:9: error: ‘CmdShowType’ undeclared here (not in a function)
  283 |         CmdShowType
      |         ^~~~~~~~~~~
dbgsh.c:283:9: warning: excess elements in struct initializer
dbgsh.c:283:9: note: (near initialization for ‘ShowCmds[16]’)
dbgsh.c:369:22: error: unknown type name ‘Collection’
  369 | static void ExecCmd (Collection* Args, const CmdEntry* Tab, unsigned Count)
      |                      ^~~~~~~~~~
dbgsh.c:960:22: error: unknown type name ‘Collection’
  960 | static void CmdHelp (Collection* Args attribute ((unused)))
      |                      ^~~~~~~~~~
dbgsh.c:968:22: error: unknown type name ‘Collection’
  968 | static void CmdLoad (Collection* Args)
      |                      ^~~~~~~~~~
dbgsh.c:993:22: error: unknown type name ‘Collection’
  993 | static void CmdQuit (Collection* Args attribute ((unused)))
      |                      ^~~~~~~~~~
dbgsh.c:1002:22: error: unknown type name ‘Collection’
 1002 | static void CmdShow (Collection* Args)
      |                      ^~~~~~~~~~
dbgsh.c:1011:26: error: unknown type name ‘Collection’
 1011 | static void CmdShowHelp (Collection* Args attribute ((unused)))
      |                          ^~~~~~~~~~
dbgsh.c:1019:33: error: unknown type name ‘Collection’
 1019 | static void CmdShowChildScopes (Collection* Args)
      |                                 ^~~~~~~~~~
dbgsh.c:1065:29: error: unknown type name ‘Collection’
 1065 | static void CmdShowCSymbol (Collection* Args)
      |                             ^~~~~~~~~~
dbgsh.c:1129:30: error: unknown type name ‘Collection’
 1129 | static void CmdShowFunction (Collection* Args)
      |                              ^~~~~~~~~~
dbgsh.c:1175:26: error: unknown type name ‘Collection’
 1175 | static void CmdShowLine (Collection* Args)
      |                          ^~~~~~~~~~
dbgsh.c:1237:29: error: unknown type name ‘Collection’
 1237 | static void CmdShowLibrary (Collection* Args)
      |                             ^~~~~~~~~~
dbgsh.c:1297:28: error: unknown type name ‘Collection’
 1297 | static void CmdShowModule (Collection* Args)
      |                            ^~~~~~~~~~
dbgsh.c:1358:27: error: unknown type name ‘Collection’
 1358 | static void CmdShowScope (Collection* Args)
      |                           ^~~~~~~~~~
dbgsh.c:1421:29: error: unknown type name ‘Collection’
 1421 | static void CmdShowSegment (Collection* Args)
      |                             ^~~~~~~~~~
dbgsh.c:1481:28: error: unknown type name ‘Collection’
 1481 | static void CmdShowSource (Collection* Args)
      |                            ^~~~~~~~~~
dbgsh.c:1544:26: error: unknown type name ‘Collection’
 1544 | static void CmdShowSpan (Collection* Args)
      |                          ^~~~~~~~~~
dbgsh.c:1596:28: error: unknown type name ‘Collection’
 1596 | static void CmdShowSymbol (Collection* Args)
      |                            ^~~~~~~~~~
dbgsh.c:1647:28: error: unknown type name ‘Collection’
 1647 | static void CmdShowSymDef (Collection* Args)
      |                            ^~~~~~~~~~
dbgsh.c:1694:28: error: unknown type name ‘Collection’
 1694 | static void CmdShowSymRef (Collection* Args)
      |                            ^~~~~~~~~~
dbgsh.c:1741:26: error: unknown type name ‘Collection’
 1741 | static void CmdShowType (Collection* Args)
      |                          ^~~~~~~~~~
dbgsh.c:1798:24: error: unknown type name ‘Collection’
 1798 | static void CmdUnload (Collection* Args attribute ((unused)))
      |                        ^~~~~~~~~~
dbgsh.c:1812:34: error: unknown type name ‘Collection’
 1812 | static int Parse (char* CmdLine, Collection* Args)
      |                                  ^~~~~~~~~~
dbgsh.c: In function ‘main’:
dbgsh.c:1873:5: error: unknown type name ‘Collection’
 1873 |     Collection Args = STATIC_COLLECTION_INITIALIZER;
      |     ^~~~~~~~~~
dbgsh.c:1873:23: error: ‘STATIC_COLLECTION_INITIALIZER’ undeclared (first use in this function)
 1873 |     Collection Args = STATIC_COLLECTION_INITIALIZER;
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dbgsh.c:1873:23: note: each undeclared identifier is reported only once for each function it appears in
dbgsh.c:1876:5: error: implicit declaration of function ‘InitCmdLine’ [-Wimplicit-function-declaration]
 1876 |     InitCmdLine (&argc, &argv, "dbgsh");
      |     ^~~~~~~~~~~
dbgsh.c:1879:9: error: ‘ArgCount’ undeclared (first use in this function)
 1879 |     if (ArgCount > 1) {
      |         ^~~~~~~~
dbgsh.c:1882:13: error: implicit declaration of function ‘CollAppend’ [-Wimplicit-function-declaration]
 1882 |             CollAppend (&Args, ArgVec[I]);
      |             ^~~~~~~~~~
dbgsh.c:1882:32: error: ‘ArgVec’ undeclared (first use in this function); did you mean ‘argc’?
 1882 |             CollAppend (&Args, ArgVec[I]);
      |                                ^~~~~~
      |                                argc
dbgsh.c:1886:9: error: implicit declaration of function ‘ExecCmd’ [-Wimplicit-function-declaration]
 1886 |         ExecCmd (&Args, MainCmds, sizeof (MainCmds) / sizeof (MainCmds[0]));
      |         ^~~~~~~
dbgsh.c:1901:13: error: implicit declaration of function ‘Parse’ [-Wimplicit-function-declaration]
 1901 |         if (Parse (Input, &Args) == 0 || CollCount (&Args) == 0) {
      |             ^~~~~
dbgsh.c:1901:42: error: implicit declaration of function ‘CollCount’ [-Wimplicit-function-declaration]
 1901 |         if (Parse (Input, &Args) == 0 || CollCount (&Args) == 0) {
      |                                          ^~~~~~~~~
dbgsh.c:1910:5: error: implicit declaration of function ‘DoneCollection’ [-Wimplicit-function-declaration]
 1910 |     DoneCollection (&Args);
      |     ^~~~~~~~~~~~~~
</pre>

--- a/.github/workflows/build-on-pull-request.yml
+++ b/.github/workflows/build-on-pull-request.yml
@@ -27,6 +27,9 @@ jobs:
       - name: Build the tools.
         shell: bash
         run: make -j2 bin USER_CFLAGS=-Werror
+      - name: Build the dbginfo example
+        shell: bash
+        run: make -j2 src dbginfo
       - name: Build the utilities.
         shell: bash
         run: make -j2 util

--- a/.github/workflows/build-on-pull-request.yml
+++ b/.github/workflows/build-on-pull-request.yml
@@ -29,7 +29,7 @@ jobs:
         run: make -j2 bin USER_CFLAGS=-Werror
       - name: Build the dbginfo example
         shell: bash
-        run: make -j2 -C src dbginfo
+        run: make -j2 -C src test
       - name: Build the utilities.
         shell: bash
         run: make -j2 util

--- a/.github/workflows/build-on-pull-request.yml
+++ b/.github/workflows/build-on-pull-request.yml
@@ -29,7 +29,7 @@ jobs:
         run: make -j2 bin USER_CFLAGS=-Werror
       - name: Build the dbginfo example
         shell: bash
-        run: make -j2 src dbginfo
+        run: make -j2 -C src dbginfo
       - name: Build the utilities.
         shell: bash
         run: make -j2 util

--- a/Makefile
+++ b/Makefile
@@ -50,7 +50,7 @@ test:
 # GNU "check" target, which runs all tests
 check:
 	@$(MAKE) -C .github/checks checkstyle --no-print-directory
-	@$(MAKE) -C src dbginfo               --no-print-directory
+	@$(MAKE) -C src test                  --no-print-directory
 	@$(MAKE) test
 	@$(MAKE) -C targettest platforms      --no-print-directory
 	@$(MAKE) -C samples platforms         --no-print-directory

--- a/Makefile
+++ b/Makefile
@@ -50,6 +50,7 @@ test:
 # GNU "check" target, which runs all tests
 check:
 	@$(MAKE) -C .github/checks checkstyle --no-print-directory
+	@$(MAKE) -C src dbginfo               --no-print-directory
 	@$(MAKE) test
 	@$(MAKE) -C targettest platforms      --no-print-directory
 	@$(MAKE) -C samples platforms         --no-print-directory

--- a/src/Makefile
+++ b/src/Makefile
@@ -21,7 +21,7 @@ PROGS = ar65     \
         sim65    \
         sp65
 
-.PHONY: all mostlyclean clean install zip avail unavail bin $(PROGS) dbginfo
+.PHONY: all mostlyclean clean install zip avail unavail bin $(PROGS)
 
 .SUFFIXES:
 
@@ -168,7 +168,19 @@ $(eval $(call OBJS_template,common))
 
 $(foreach prog,$(PROGS),$(eval $(call PROG_template,$(prog))))
 
-dbginfo:
-$(eval $(call PROG_template,dbginfo))
+
+.PHONY: dbginfo dbgsh test
+
+test: dbginfo dbgsh
+
+$(eval $(call OBJS_template,dbginfo))
+
+dbginfo: $(dbginfo_OBJS)
+
+../wrk/dbgsh$(EXE_SUFFIX): $(dbginfo_OBJS) ../wrk/common/common.a
+	$(CC) $(LDFLAGS) -o $@ $^ $(LDLIBS)
+
+dbgsh: ../wrk/dbgsh$(EXE_SUFFIX)
+
 
 -include $(DEPS)

--- a/src/Makefile
+++ b/src/Makefile
@@ -21,7 +21,7 @@ PROGS = ar65     \
         sim65    \
         sp65
 
-.PHONY: all mostlyclean clean install zip avail unavail bin $(PROGS)
+.PHONY: all mostlyclean clean install zip avail unavail bin $(PROGS) dbginfo
 
 .SUFFIXES:
 
@@ -167,5 +167,8 @@ $(eval $(call OBJS_template,common))
 	$(AR) r $@ $?
 
 $(foreach prog,$(PROGS),$(eval $(call PROG_template,$(prog))))
+
+dbginfo:
+$(eval $(call PROG_template,dbginfo))
 
 -include $(DEPS)

